src/analyz/to_json.rs

```diff
@@ -107,22 +108,43 @@ impl<'tcx> TraitInst<'tcx> {
     ) -> TraitInst<'tcx> {
         let ex_trait_ref = ty::ExistentialTraitRef::erase_self_ty(tcx, trait_ref);
 
+        let mut all_super_traits = HashSet::new();
+        let mut pending = vec![trait_ref.def_id];
+        all_super_traits.insert(trait_ref.def_id);
+        while let Some(def_id) = pending.pop() {
+            let super_preds = tcx.super_predicates_of(def_id);
+            for &(ref pred, _) in super_preds.predicates {
+                let tpred = match tcx.erase_late_bound_regions(pred.kind()) {
+                    ty::PredicateKind::Clause(ty::Clause::Trait(x)) => x,
+                    ty::PredicateKind::Clause(ty::Clause::TypeOutlives(..)) => continue,
+                    _ => panic!("unexpected predicate kind: {:?}", pred),
+                };
+                assert_eq!(tpred.polarity, ty::ImplPolarity::Positive);
+                if all_super_traits.insert(tpred.trait_ref.def_id) {
+                    pending.push(tpred.trait_ref.def_id);
+                }
+            }
+        }
+        let mut all_super_traits = all_super_traits.into_iter().collect::<Vec<_>>();
+        all_super_traits.sort();
+
         let mut projs = Vec::new();
-        // FIXME: build projs for supertrait trait_refs as well
-        for ai in tcx.associated_items(trait_ref.def_id).in_definition_order() {
-            match ai.kind {
-                ty::AssocKind::Type | ty::AssocKind::OpaqueTy => {},
-                _ => continue,
+        for super_trait_def_id in all_super_traits {
+            for ai in tcx.associated_items(super_trait_def_id).in_definition_order() {
+                match ai.kind {
+                    ty::AssocKind::Type => {},
+                    _ => continue,
+                }
+                let proj_ty = tcx.mk_projection(ai.def_id, trait_ref.substs);
+                let actual_ty = tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), proj_ty);
+                projs.push(ty::ExistentialProjection {
+                    def_id: ai.def_id,
+                    substs: ex_trait_ref.substs,
+                    term: actual_ty.into(), // TO REVIEW: structure of Term has changed
+                });
```

Comment:
What does this "to review" comment refer to?

Reply:
I'm honestly not sure. This came from https://github.com/GaloisInc/mir-json/commit/c13957d2d3017f28a0299d03b385cac88436de4d, so perhaps @m10f could offer some insight here.

Reply:
I spoke to @m10f. This comment is referring to the fact that this `rust` commit (https://github.com/rust-lang/rust/commit/79db32b64ec946afbef2ed50282f03b71d933480) changed `Term` from an `enum`:

```rs
pub enum Term<'tcx> {
    Ty(Ty<'tcx>),
    Const(Const<'tcx>),
}
```

To a struct:

```rs
pub struct Term<'tcx> {
    ptr: NonZeroUsize,
    marker: PhantomData<(Ty<'tcx>, Const<'tcx>)>,
}
```

And this comment served as a reminder to check if changing this line in `mir-json` from `term: ty::Term::Ty(actual_ty)` to `term: actual_ty.into()` is the proper way to migrate the code. That being said, I took a closer look at the `rust` commit, and I am fairly certain that this is correct. That commit _also_ migrates a similar bit of code [here](https://github.com/rust-lang/rust/commit/79db32b64ec946afbef2ed50282f03b71d933480#diff-ea6f9d55e90333a1f5fc5e202d46862a6a031410831f2df42317ac024f821162R352) in the exact same way:

```diff
-                    term: Term::Ty(output_ty),
+                    term: output_ty.into(),
```

As such, I think the current code is fine. I will remove the `TO REVIEW` comment and land this, but feel free to object if I've overlooked something.

Reply:
Yes, that sounds like the correct fix.  The new `Term` is a sort of "packed enum" that uses (I assume) pointer tagging instead of a separate discriminant field to distinguish the `Ty` and `Const` cases, so it's only one word wide instead of two.